### ISSUE TITLE
Set LINK_LIBRARIES_ONLY_TARGETS=TRUE by default and verify Trilinos is clean with CMake 3.23+

### ISSUE THREAD

**Issue by bartlettroscoe** [trilinos/Trilinos#10081]
As discussed in https://github.com/TriBITSPub/TriBITS/issues/299#issuecomment-1012312724, CMake 3.23 will learn the option `LINK_LIBRARIES_ONLY_TARGETS` that if set to `TRUE`, will help catch errors when linking against undefined targets.  With the updated TriBITS, raw library names for external packages/TPLs are no longer directly passed to `target_link_libraries()` and therefore, unless people are directly calling `target_link_libraries()` against raw lib names in Trilinos CMakeLists.txt files, we should be able to set `LINK_LIBRARIES_ONLY_TARGETS` to true with Trilinos.

But before I felt confident about that, Trilinos would need to upgrade its minimum CMake version to CMake 3.23+ and all of the Trilinos PR and ATDM Trilinos builds would need to be tested with this option enabled with the upgraded CMake.

**Comment by bartlettroscoe** [comment]
@sebrowne, now that Trilinos is requiring CMake 3.23+, we can pull the trigger on this one.  This will make the CMakeLists.txt files more robust and will be critical when refactoring to use raw library linking as listed in:

* https://github.com/TriBITSPub/TriBITS/issues/569

**Comment by sebrowne** [comment]
So to clarify, this will turn "nonexistant" targets called with target_link_libaries(myTarget <TYPE> someBadTarget) into configure errors instead of doing "lld: error: -lsomeBadTarget not found" at link time?

**Comment by bartlettroscoe** [comment]
> So to clarify, this will turn "nonexistant" targets called with `target_link_libaries(myTarget <TYPE> someBadTarget)` into configure errors instead of doing `lld: error: -lsomeBadTarget not found` at link time?

@sebrowne, yes.  From [LINK_LIBRARIES_ONLY_TARGETS](https://cmake.org/cmake/help/v3.23/prop_tgt/LINK_LIBRARIES_ONLY_TARGETS.html?highlight=link_libraries_only_targets):

> Enforce that link items that can be target names are actually existing targets.

Otherwise, if someone makes a mistake, you get a build-time error about a missing dependency (which is confusing for some people to figure out).  That does not happen much with current TriBITS due to auto-linking between dependent package libraries but if Trilinos moves to manual linking (see https://github.com/TriBITSPub/TriBITS/issues/569) then this will happen all of the time.

**Comment by bartlettroscoe** [comment]
FYI: The recent refactoring of the TriBITS legacy TPL system that creates proper CMake IMPORTED targets will allow Trilinos (and TriBITS) to set `LINK_LIBRARIES_ONLY_TARGETS` to `TRUE` by default and that will not break Trilinos configure scripts that are passing in raw library names.

**Comment by bartlettroscoe** [comment]
@sebrowne, @ccober6, and @jwillenbring, this is another one that should be turned on for Trilinos 15.0.

**Comment by sebrowne** [comment]
We'll get on it immediately (I'll write a small story and have the team take care of it prior to the release).  Thanks for circling back!

**Comment by bartlettroscoe** [comment]
@sebrowne, FYI, I had to revert the merge of the PR:

 * #12362

with the revert PR:

* #12372

because it was breaking the installed `ZoltanConfig.cmake` file, see:

* #12371

First, we need to beef up the test `TrilinosInstallTests_find_package_Trilinos` defined at:

https://github.com/trilinos/Trilinos/blob/b3965b6a93763aa5c39f02e0e77aca188a24f6ef/packages/TrilinosInstallTests/CMakeLists.txt#L239

and:

https://github.com/trilinos/Trilinos/blob/b3965b6a93763aa5c39f02e0e77aca188a24f6ef/packages/TrilinosInstallTests/find_package_Trilinos/CMakeLists.txt#L1-L11

to try to expose this failure.  I am surprised that test passed in all of the PR builds shown [here](https://trilinos-cdash.sandia.gov/queryTests.php?project=Trilinos&date=2023-10-05&filtercount=2&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=PR-12362-&field2=testname&compare2=61&value2=TrilinosInstallTests_find_package_Trilinos) before it was merged

We should try adding a dummy executable as part of that dummy CMake project [find_package_Trilinos](https://github.com/trilinos/Trilinos/tree/develop/packages/TrilinosInstallTests/find_package_Trilinos) and linking in the Trilinos libraries and see if it triggers that error reported in  #12371.

As for setting `CMAKE_LINK_LIBRARIES_ONLY_TARGETS=ON`, there are a couple of options we can consider off the top of my head:

**Option-1**: Set the property `LINK_LIBRARIES_ONLY_TARGETS=OFF` just for those two targets that are linking in raw `m`.

**Option-2**: Create a new Trilinos TriBITS TPL `MLib` with the trivial file `FindTPLMLib.cmake` that defines the `MLib` TPL to just be `TPL_MLib_LIBRARIES=m` and then make Zoltan and SuiteSparse depend on `MLib`.

Option-1 involves less code but it seems like kind of a hack.  Option-2 involves a little more overhead (in a new file `FindTPLMLib.cmake` and installing a file `MLibConfig.cmake` that has to be included in downstream `<Package>Config.cmake` files) but it very clean consistent with how the rest of TriBITS TPLs work.

NOTE: One option **not** to try to export the created imported target because CMake does not support that :-(

**Comment by sebrowne** [comment]
@bartlettroscoe I added that case to the test and tested it, it does work and catch the issue.

**Comment by bartlettroscoe** [comment]
For now, we can put this back into the backlog since this is not worth fixing for now.  If we get Trilinos internal packages using raw CMake, then we may want to try harder to turn this on and address the issue with the `m` library.
